src/vvwe.rs

```diff
@@ -0,0 +1,256 @@
+// Causality barrier
+// Keeps for each known peer, keeps track of the latest clock seen
+// And a set of messages that are from the future
+// and outputs the full-in-order sequence of messages
+//
+//
+
+#![allow(missing_docs)]
+
+use std::collections::*;
+use serde::{Deserialize, self, Serialize};
+
+/// Version Vector with Exceptions
+#[derive(Debug, Serialize, Deserialize)]
+pub struct CausalityBarrier<A: Actor, T: CausalOp<A>> {
+    peers: HashMap<A, VectorEntry>,
+    local_id: A,
+    pub buffer: HashMap<Dot<A>, T>,
+}
+
+type LogTime = u64;
+
+#[derive(Serialize, Deserialize, Clone, Debug)]
+pub struct VectorEntry {
+    // The version of the next message we'd like to see
+    max_version: LogTime,
+    exceptions: HashSet<LogTime>,
+}
+
+
+impl VectorEntry {
+    pub fn new() -> Self {
+        VectorEntry { max_version: 0, exceptions: HashSet::new() }
+    }
+
+    pub fn increment(&mut self, clk: LogTime) {
+        // We've just found an exception
+        if clk < self.max_version {
+            self.exceptions.take(&clk);
+        } else if clk == self.max_version {
+            self.max_version = self.max_version + 1;
+        } else {
+            let mut x = self.max_version + 1;
+            while x < clk {
+                self.exceptions.insert(x);
+                x = x + 1;
+            }
+        }
+    }
+
+    pub fn is_ready(&self, clk: &LogTime) -> bool {
+        *clk < self.max_version && !self.exceptions.contains(clk)
+    }
+
+    /// Calculate the difference between a remote VectorEntry and ours.
+    /// Specifically, we want the set of operations we've seen that the remote hasn't
+    pub fn diff_from(&self, other: &Self) -> HashSet<LogTime> {
+        // 1. Find (new) operations that we've seen locally that the remote hasn't
+        let local_ops = (other.max_version.into()..self.max_version.into()).into_iter().filter(|ix : &u64| {
+            !self.exceptions.contains(&(*ix).into())
+        }).map(LogTime::from);
+
+        // 2. Find exceptions that we've seen.
+        let mut local_exceptions  = other.exceptions.difference(&self.exceptions).map(|ix| ix.to_owned());
+
+        local_ops.chain(&mut local_exceptions).collect()
+    }
+}
+
+use crate::vclock::{Actor, Dot};
+
+pub trait CausalOp<A> {
```

Comment:
implementing this trait for different kinds of `Op` types allows this barrier to work with multiple CRDTs